prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +26,131 @@ def floatify(value, default=float("nan")):
-        matching_subs = substation_groupby.get_group(lookup_params)
-    except KeyError:
-        # If this ZIP code does not contain substations, this block will execute, and
-        # we select a set of 'nearby' substations
-        zip_range = [int(generator.loc["ZIP"]) + offset for offset in range(-100, 101)]
-        zip_range_strings = [str(z).rjust(5, "0") for z in zip_range]
-        try:
-            matching_subs = pd.concat(
-                [
-                    substation_groupby.get_group((generator.loc["interconnect"], z))
-                    for z in zip_range_strings
-                    if (generator.loc["interconnect"], z) in substation_groupby.groups
-                ]
-            )
-        except ValueError:
-            # If no matching subs within the given interconnection and ZIPs, give up
+
+    def get_closest_substation(generator, state_trees, subs_state_lookup):
+        if not isinstance(generator["xyz"], list):
             return pd.NA
-    distance_to_subs = matching_subs.apply(
-        lambda x: haversine((x.lat, x.lon), (generator.lat, generator.lon)),
+        if pd.isnull(generator["voltage_class"]) or generator["Pmax"] < 100:
+            grouper_key = generator["interconnect"]
+        else:
+            grouper_key = (generator["interconnect"], generator["voltage_class"])
+        chord_dist, array_index = voltage_trees[grouper_key].query(generator["xyz"])
+        sub_id = subs_voltage_lookup[grouper_key][array_index]
+        # Translate chord distance (unit circle) to great circle distance (miles)
+        dist_in_miles = 3963 * 2 * asin(chord_dist / 2)  # use 3963 mi as earth radius
+        return pd.Series({"dist": dist_in_miles, "sub_id": sub_id})
+
+    def classify_voltages(voltage, voltage_ranges):
+        for v_range, bounds in voltage_ranges.items():
+            if bounds["min"] <= voltage <= bounds["max"]:
+                return v_range
+        return float("nan")
+
+    voltage_ranges = {
+        "under 100": {"min": 0, "max": 99},
+        "100-161": {"min": 100, "max": 161},
+        "220-287": {"min": 220, "max": 287},
+        "345": {"min": 345, "max": 345},
+        "500": {"min": 500, "max": 500},
+        "735 and above": {"min": 735, "max": float("inf")},
+    }
+
+    # Translate lat/lon to 3D positions (assume spherical earth, origin at center)
+    substations_with_xyz = substations.assign(
+        xyz=substations.apply(lambda x: latlon_to_xyz(x["lat"], x["lon"]), axis=1)
+    )
+    generators_with_xyz = generators.assign(
+        xyz=generators.apply(
+            lambda x: (
+                pd.NA
+                if pd.isna(x["lat"]) or pd.isna(x["lon"])
+                else latlon_to_xyz(x["lat"], x["lon"])
+            ),
+            axis=1,
+        )
+    )
+
+    # Bin voltages into broad classes
+    substations_with_xyz["voltage_class"] = substations["MAX_VOLT"].map(
+        lambda x: classify_voltages(x, voltage_ranges)
```

Comment:
Is the column `voltage_class` in data frame `substations_with_xyz` used anywhere?

Reply:
It appears not, I'll remove this and double-check that everything still works.

Reply:
Done.